src/connector.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const net = require('net');
+const lookupAll = require('dns-lookup-all');
+
+class Connector {
+  constructor(options, multiSubnetFailover) {
+    this.options = options;
+    this.multiSubnetFailover = multiSubnetFailover;
+  }
+
+  execute(cb) {
+    if (net.isIP(this.options.host)) {
+      this.executeForIP(cb);
+    } else {
+      this.executeForHostname(cb);
+    }
+  }
+
+  executeForIP(cb) {
+    const socket = net.connect(this.options);
+
+    const onError = (err) => {
+      socket.removeListener('error', onError);
+      socket.removeListener('connect', onConnect);
+
+      socket.destroy();
+
+      cb(err);
+    };
+
+    const onConnect = () => {
+      socket.removeListener('error', onError);
+      socket.removeListener('connect', onConnect);
+
+      cb(null, socket);
+    };
+
+    socket.on('error', onError);
+    socket.on('connect', onConnect);
+  }
+
+  executeForHostname(cb) {
+    lookupAll(this.options.host, 4, (err, addresses) => {
```

Comment:
> I'm wondering - can we just drop all of the sequential connection handling and just do parallel connections, always? Do you see any disadvantages to that?

Parallel connections seem to have cause a few issues before with services getting overwhelmed on startup when they try to make a ton of parallel connections.

Oe strategy that has been adopted is to first try just the the first IP address returned by DNS lookup, and only if that fails, then attempt the parallel connections Now you have to slice the connection timeout between the first attempt and the second attempt. This slicing of the timeout between the first attempt and the rest, caused a different problem, in that it increased the likely of connection timeout. The team is now investigating yet another tweak, which is to do the slicing for local databases but not for Azure databases (or something close to that).

Summary is that, it can get complicated :) Sequential as the default and Parallel as opt-in sounds to me like a reasonable place to start for this feature.

> Also, I think we should look into the instance listener connection code, I don't think that was updated for support of parallel connections.

What's "instance listener connection code"? Is this required to merge this PR?